### ISSUE TITLE
Docs: more metadata for each rule

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#4156]
### Suggested Changes

The docs is hard to navigate without search.

Regarding search, I don't know if there're plans to integrate Algolia—it's going to be free and pretty convenient to set up😉

This doc request is slightly more than that. What I'd want to have is more metadata about each rule: besides "fixable", "recommended" etc., we can also have "formatting", "possible errors", etc., and "extending ESLint" for those rules that require you to turn off the ESLint counterpart.

We should consider adding these metadata in front matter besides as a checklist. The front matter will automatically become metadata that can be handled as structured data by either a Docusaurus plugin or theme components. I envision something like this:

```md
---
extend_eslint: true
category: formatting
recommended: false
fixable: true
type_information: false
---

# Require or disallow an empty line between class members (`lines-between-class-members`)
```

When we do have that pulled up, we are in a good position to do two things:

1. Generate a better landing page automatically by rendering all the metadata in a table; (we can keep the current README file for GitHub interface, but it would be better to have something fancier on the website)
2. Allow filtering rules by applying conditions, e.g. `category=suggestions&recommended=false`. ref. the Docusaurus showcase UI: https://docusaurus.io/showcase

While Docusaurus does not provide a GraphQL layer like Gatsby (https://github.com/facebook/docusaurus/issues/4492) where we can dynamically query all front matter, we can still enhance the docs plugin and propagate the front matter to the landing page. If there's interest in this, I can offer some support.

### Affected URL(s)

https://typescript-eslint.io/rules/

**Comment by JoshuaKGoldberg** [comment]
Hey @Josh-Cena! Strong +1 and I appreciate you writing out these points in an issue! Agreed all around smile 😄. 

The website is moving pretty fast and we haven't gotten around to filing issues for quite everything, so apologies for not surfacing that we do have some plans pending or in motion. You can see the raw notes here: https://github.com/orgs/typescript-eslint/projects/1/views/1.

In order:

> I don't know if there're plans to integrate Algolia

🙌 Yes! I got the onboarding flow started last week and they _just_ sent an email. Hopefully we'll have it set up Soon™️!

> more metadata about each rule: besides "fixable", "recommended" etc., we can also have ...

Super interesting. Deferring to @armano2 post #4114. Whatever set of extra metadata is included, it should probably live in the `metadata` field on the rule and have similar linting/tests to stay up-to-date with the rules docs.

> "formatting"

Yeah, and we should add a _"ugh just use Prettier please"_ notice to those rules' pages 😉 

> "possible errors"

Is this different from the existing `messageIds` on each rule?

**Comment by Josh-Cena** [comment]
"formatting" and "possible errors" are categories ESLint uses: https://eslint.org/docs/rules/ (and excuse me, it's "Possible Problems"🤦‍♂️) And no, I'm not referring to `messageIds`

**Comment by armano2** [comment]
> 
> 
> Hey @Josh-Cena! Strong +1 and I appreciate you writing out these points in an issue! Agreed all around smile 😄.
> 
> The website is moving pretty fast and we haven't gotten around to filing issues for quite everything, so apologies for not surfacing that we do have some plans pending or in motion. You can see the raw notes here: https://github.com/orgs/typescript-eslint/projects/1/views/1.
> 
> In order:
> 
> > I don't know if there're plans to integrate Algolia
> 
> 🙌 Yes! I got the onboarding flow started last week and they _just_ sent an email. Hopefully we'll have it set up Soon™️!
> 
> > more metadata about each rule: besides "fixable", "recommended" etc., we can also have ...
> 
> Super interesting. Deferring to @armano2 post #4114. Whatever set of extra metadata is included, it should probably live in the `metadata` field on the rule and have similar linting/tests to stay up-to-date with the rules docs.
> 
> > "formatting"
> 
> Yeah, and we should add a _"ugh just use Prettier please"_ notice to those rules' pages 😉
> 
> > "possible errors"
> 
> Is this different from the existing `messageIds` on each rule?

we can use docs field, and some custom fields in there (per rule)

```
meta: {
    docs: {
        categories: ['essential', 'etc'],
```

something similar to what has been done in here: https://github.com/armano2/eslint-plugin-vue/blob/4261a396860a1739c652c0a0d59d79b80d8b6a6a/lib/rules/require-default-prop.js#L28

**Comment by bradzacher** [comment]
> we can use docs field, and some custom fields in there

ESLint core actually removed all of the categories from their rule meta because it's something additional to maintain - things drift over time - it just wasn't worth the effort.

> [The meta.docs.category property has been removed from all core rules.](https://eslint.org/blog/2021/10/eslint-v8.0.0-released)

Instead now they group rules based on `rule.meta.type` (which is the same way the CLI has always worked). I think that it's a good philosophy to have. No doubt we'd have the same problem here - narrow down the configuration points to make it easier to keep things in sync and up-to-date.

For reference the three "types" are:
- `problem` - "Possible Problems" - "These rules relate to possible logic errors in code"
- `suggestion` - "Suggestions" - "These rules suggest alternate ways of doing things"
- `layout` - "Layout & Formatting" - "These rules care about how the code looks rather than how it executes"

> something similar to what has been done in vuejs/eslint-plugin-vue

Worth noting that the vue plugin also triples down on the categorisation aspect and it generates many, many different configs based on the config.
I personally don't hugely want to offer too many configs to the world because it's historically been a big enough pain to just manage the recommended set against users' expectations (a lot of people are really opinionated and are LOUD about it). Vue is a bit different to us because it's much smaller than us and (from what I've seen) the community is substantially more quiet and less involved.

**Comment by JoshuaKGoldberg** [comment]
Splitting to keep this issue's tracking to just the rule metadata discussion:
* #4165
* #4168 _(not significantly discussed here, but related)_

**Comment by Josh-Cena** [comment]
So here's my plan:

1. Add front matter to each rule doc.
2. [Wrap the default Doc item component](https://docusaurus.io/docs/next/using-themes#wrapping-theme-components) to display the rule attributes in a checklist format similar to the current one.
3. Enhance the docs plugin for the rules and send the rule metadata as props to the rule overview page so that it can be filtered/queried in the table.

**Comment by JoshuaKGoldberg** [comment]
Sounds great @Josh-Cena!

**Comment by Josh-Cena** [comment]
Lots of things have happened. I like the rule table now
